File: editorial-intelligence/workers/scheduler.py
"""Scheduler (Phase 6) — decides whether a run is due, for
manual/hourly/daily/weekly modes.

No real cron/timer — a pure function of "mode + last_run_at + now",
matching the spec's "Worker phải mock hoàn toàn. Không network."
"""
import datetime
from typing import Optional

_INTERVALS = {
    "hourly": datetime.timedelta(hours=1),
    "daily": datetime.timedelta(days=1),
    "weekly": datetime.timedelta(weeks=1),
}
_MODES = ("manual", "hourly", "daily", "weekly")


class Scheduler:
    def __init__(self, mode: str = "manual"):
        if mode not in _MODES:
            raise ValueError(f"Unknown schedule mode: {mode!r} (expected one of {_MODES})")
        self._mode = mode

    @property
    def mode(self) -> str:
        return self._mode

    def is_due(self, last_run_at: Optional[str], now: Optional[datetime.datetime] = None) -> bool:
        """`manual` is always due — an explicit `editorial worker run`
        call always runs. hourly/daily/weekly are due only if no run has
        happened yet, or the interval has elapsed since the last one."""
        if self._mode == "manual":
            return True
        if last_run_at is None:
            return True
        now = now or datetime.datetime.now(datetime.timezone.utc)
        last = datetime.datetime.fromisoformat(last_run_at)
        return now - last >= _INTERVALS[self._mode]

File: workers/editorial-os/src/worker/scheduler.ts
/**
 * Scheduler — 1:1 port of editorial-intelligence/workers/scheduler.py.
 *
 * Kept for rule-fidelity with Python's manual/hourly/daily/weekly
 * vocabulary and its `isDue()` gate — in the Phase 8 deployment, the
 * actual firing schedule is Cloudflare's own Cron Trigger (every 30
 * minutes, see the `crons` entry in wrangler.toml), so this class's role here is the
 * same defensive check Python's `WorkerRunner.run()` already performed
 * before doing any work, not a replacement for the Cron Trigger itself.
 */

export type ScheduleMode = "manual" | "hourly" | "daily" | "weekly";

const INTERVALS_MS: Record<Exclude<ScheduleMode, "manual">, number> = {
  hourly: 60 * 60 * 1000,
  daily: 24 * 60 * 60 * 1000,
  weekly: 7 * 24 * 60 * 60 * 1000,
};

const MODES: ScheduleMode[] = ["manual", "hourly", "daily", "weekly"];

export class Scheduler {
  readonly mode: ScheduleMode;

  constructor(mode: ScheduleMode = "manual") {
    if (!MODES.includes(mode)) {
      throw new Error(`Unknown schedule mode: ${JSON.stringify(mode)} (expected one of ${MODES.join(", ")})`);
    }
    this.mode = mode;
  }

  /** `manual` is always due. hourly/daily/weekly are due only if no run
   * has happened yet, or the interval has elapsed since the last one. */
  isDue(lastRunAt: string | null, now: Date = new Date()): boolean {
    if (this.mode === "manual") return true;
    if (lastRunAt === null) return true;
    const last = Date.parse(lastRunAt);
    return now.getTime() - last >= INTERVALS_MS[this.mode];
  }
}
